benchto-driver/src/main/java/com/teradata/benchto/driver/listeners/BenchmarkServiceExecutionListener.java

```diff
@@ -94,34 +105,47 @@ public void executionStarted(QueryExecution execution)
-
-        if (executionResult.getPrestoQueryId().isPresent()) {
-            requestBuilder.addAttribute("prestoQueryId", executionResult.getPrestoQueryId().get());
-        }
-
-        if (!executionResult.isSuccessful()) {
-            requestBuilder.addAttribute("failureMessage", executionResult.getFailureCause().getMessage());
-            requestBuilder.addAttribute("failureStackTrace", stackTraceToString(executionResult));
-
-            if (executionResult.getFailureCause() instanceof SQLException) {
-                requestBuilder.addAttribute("failureSQLErrorCode", "" + ((SQLException) executionResult.getFailureCause()).getErrorCode());
-            }
-        }
-
-        benchmarkServiceClient.finishExecution(executionResult.getBenchmark().getUniqueName(), executionResult.getBenchmark().getSequenceId(),
-                executionSequenceId(executionResult.getQueryExecution()), requestBuilder.build());
+        getMeasurements(executionResult)
+                .thenApply(measurements -> {
+                    FinishRequestBuilder requestBuilder = new FinishRequestBuilder()
+                            .withStatus(executionResult.isSuccessful() ? ENDED : FAILED)
+                            .addMeasurements(measurements);
+
+                    if (executionResult.getPrestoQueryId().isPresent()) {
+                        requestBuilder.addAttribute("prestoQueryId", executionResult.getPrestoQueryId().get());
+                    }
+
+                    if (!executionResult.isSuccessful()) {
+                        requestBuilder.addAttribute("failureMessage", executionResult.getFailureCause().getMessage());
+                        requestBuilder.addAttribute("failureStackTrace", stackTraceToString(executionResult));
+
+                        if (executionResult.getFailureCause() instanceof SQLException) {
+                            requestBuilder.addAttribute("failureSQLErrorCode", "" + ((SQLException) executionResult.getFailureCause()).getErrorCode());
+                        }
+                    }
+                    return requestBuilder.build();
+                })
+                .thenAccept(request -> {
+                    benchmarkServiceClient.finishExecution(
+                            executionResult.getBenchmark().getUniqueName(),
+                            executionResult.getBenchmark().getSequenceId(),
+                            executionSequenceId(executionResult.getQueryExecution()),
+                            request);
+                });
     }
 
-    private List<Measurement> getMeasurements(Measurable measurable)
+    private CompletableFuture<List<Measurement>> getMeasurements(Measurable measurable)
     {
-        ImmutableList.Builder<Measurement> measurementsList = ImmutableList.builder();
+        List<CompletableFuture<?>> providerFutures = new ArrayList<>();
+        List<Measurement> measurementsList = Collections.synchronizedList(new ArrayList<>());
```

Comment:
Yup, but `CompletableFuture.allOf` returns future of  `Void` rather than `List<T>`.